### ISSUE TITLE
Using CF_Preview_Play on multiple sounds

### ISSUE THREAD

**Issue by dsyrock** [reaper-oss/sws#1804]
I wrote a script to play two sounds
`
path1='e:\\sound1.wav'
path2='e:\\sound2.wav'
local sr1 = reaper.PCM_Source_CreateFromFile(path1)
local pv1 = reaper.CF_CreatePreview(sr1)
local sr2 = reaper.PCM_Source_CreateFromFile(path2)
local pv2 = reaper.CF_CreatePreview(sr2)

local time = reaper.time_precise()
reaper.CF_Preview_Play(pv1)
function loop()
    if reaper.time_precise()-time>=1 then
        reaper.CF_Preview_Play(pv2)
    else
        reaper.defer(loop)
    end
end
loop()
`

In this code, sound2 is unable to play. But if I change
`reaper.CF_Preview_Play(pv2)`
to
`reaper.CF_Preview_Play(reaper.CF_CreatePreview(sr2))`

Sound2 is played as expected.

So what's the different between them?

**Comment by cfillion** [comment]
The preview objects are garbage-collected if they are not used. As per the documentation:

> The preview object is automatically destroyed at the end of a defer cycle if at least one of these conditions are met:
> - playback finished
> - playback was not started using CF_Preview_Play
> - the output track no longer exists

**Comment by dsyrock** [comment]
> The preview objects are garbage-collected if they are not used to prevent any memory leaks. As per the documentation:
> 
> > The preview object is automatically destroyed at the end of a defer cycle if at least one of these conditions are met:
> > 
> > * playback finished
> > * playback was not started using CF_Preview_Play
> > * the output track no longer exists

If I misunderstood, please correct me. In my example, when the program reaches the line of "loop()", it meets the condition of "at the end of a defer cycle". Because pv2 has not been played until this point, it gets garbage collected.

**Comment by cfillion** [comment]
That is correct. At the end of the very fist defer cycle (~30ms after `CreatePreview`), the preview is considered unused and is destroyed. By the time the script does `CF_PreviewPlay` a whole second later it's long gone.

**Comment by dsyrock** [comment]
> That is correct. At the end of the next defer cycle (<=30ms after `CF_CreatePreview`), the preview is considered unused and is destroyed. By the time the script does `CF_Preview_Play` a whole second later it's long gone.


Thank you. By the way, the reaper.CF_Preview_SetValue API currently cannot set the playback duration because the length parameter is read-only. If I want to play for x seconds, I can only stop the playback by using a deferred loop to reach x seconds and then stop the playback using reaper.CF_Preview_Stop. Will you consider making the length parameter adjustable in the future official version, or a similar parameter that can limit the playback duration?

**Comment by cfillion** [comment]
Use ` CF_PCM_Source_SetSectionInfo` to limit the length of the `PCM_source`.

**Comment by dsyrock** [comment]
Big help, thanks!

**Comment by dsyrock** [comment]
Hi, cfillion, sorry to bother you again. I encountered an issue while using CF_PCM_Source_SetSectionInfo. According to the documentation, the first and second parameters are the section and PCMsource, where the section is obtained from PCM_Source_CreateFromType. However, I'm unsure how to connect the section with the source created from the audio file. My code looks like this:
```
path1='e:\\sound1.wav'
local sr1 = reaper.PCM_Source_CreateFromFile(path1)
local section = reaper.PCM_Source_CreateFromType('WAVE')
reaper.CF_PCM_Source_SetSectionInfo(section, sr1, 0, 1, false)
local pv1 = reaper.CF_CreatePreview(sr1)
reaper.PCM_Source_Destroy(sr1)
reaper.CF_Preview_Play(pv1)
```
I want to limit the playback duration of the sound1 file to 1 second, but the code doesn't seem to be working. Can you give me an example?

**Comment by dsyrock** [comment]
> ```lua
> local source  = reaper.PCM_Source_CreateFromFile(path1)
> local section = reaper.PCM_Source_CreateFromType('SECTION')
> reaper.CF_PCM_Source_SetSectionInfo(section, source, 0, 1, false)
> reaper.PCM_Source_Destroy(source)
> local preview = reaper.CF_CreatePreview(section)
> reaper.PCM_Source_Destroy(section)
> ```

Awesome!

**Comment by dsyrock** [comment]
CF_Preview_Stop and CF_Preview_StopAll generally work fine in most situations, immediately terminating the content playing in the background. However, if CF_Preview_SetValue is used to set the fadeout parameter, there seems to be a delay when stopping the playback. The background playback continues for a few seconds after the stop command is issued. I'm not sure if this duration corresponds to the fadeout length I've set. Is this behavior normal?

**Comment by cfillion** [comment]
Yes, that's the fade out feature doing its thing. Disable it (0) if you want an immediate stop without a fade.

**Comment by dsyrock** [comment]
Thanks!
